python/ray/data/_internal/datasource/iceberg_datasource.py

```diff
@@ -0,0 +1,219 @@
+                function
+            catalog_kwargs: Optional arguments to use when setting up the Iceberg
+                catalog
+        """
+        from pyiceberg.expressions import AlwaysTrue
+
+        self._scan_kwargs = scan_kwargs if scan_kwargs is not None else {}
+
+        self.table_identifier = table_identifier
+
+        self._catalog_kwargs = catalog_kwargs if catalog_kwargs is not None else {}
+
+        if "name" in self._catalog_kwargs:
+            self._catalog_name = self._catalog_kwargs.pop("name")
+        else:
+            self._catalog_name = "default"
+
+        self._row_filter = row_filter if row_filter is not None else AlwaysTrue()
+        self._selected_fields = selected_fields
+
+        if snapshot_id:
+            self._scan_kwargs["snapshot_id"] = snapshot_id
+
+        self._plan_files = None
+
+    def _get_catalog(self) -> "Catalog":
+        from pyiceberg import catalog
+
+        return catalog.load_catalog(self._catalog_name, **self._catalog_kwargs)
+
+    @property
+    def plan_files(self) -> List["FileScanTask"]:
+        """
+        Return the plan files specified by this query
+        """
+        # Calculate and cache the plan_files if they don't already exist
+        if self._plan_files is None:
+            data_scan = self._get_data_scan()
+            self._plan_files = data_scan.plan_files()
+
+        return self._plan_files
+
+    def _get_data_scan(self) -> "DataScan":
+        catalog = self._get_catalog()
+        table = catalog.load_table(self.table_identifier)
+
+        data_scan = table.scan(
+            row_filter=self._row_filter,
+            selected_fields=self._selected_fields,
+            **self._scan_kwargs,
+        )
+
+        return data_scan
+
+    def estimate_inmemory_data_size(self) -> Optional[int]:
+        # Approximate the size by using the plan files - this will not
+        # incorporate the deletes, but that's a reasonable approximation
+        # task
+        return sum(task.file.file_size_in_bytes for task in self.plan_files)
+
+    @staticmethod
+    def _distribute_tasks_into_equal_chunks(
+        plan_files: Iterable["FileScanTask"], n_chunks: int
+    ) -> List[List["FileScanTask"]]:
+        """
+        Implement a greedy knapsack algorithm to distribute the files in the scan
+        across tasks, based on their file size, as evenly as possible
+        """
+        chunks = [list() for _ in range(n_chunks)]
+        chunk_sizes = {chunk: 0 for chunk in range(n_chunks)}
+
+        # From largest to smallest, add the plan files ot the smallest chunk one at a
+        # time
+        for plan_file in sorted(plan_files, key=lambda f: f.length, reverse=True):
+            smallest_chunk = min(chunk_sizes, key=chunk_sizes.get)
+            chunks[smallest_chunk].append(plan_file)
+            # Update the size of the chunk after adding the file to it
+            chunk_sizes[smallest_chunk] += plan_file.length
```

Comment:
not super familiar with pyiceberg so this may be wrong, but should we use `file_size_in_bytes` instead of `length` for the file size, like in `estimate_inmemory_data_size()`? or are they the same?